protocol/protocol.tex

```diff
@@ -4299,30 +4808,136 @@
         \diversifier specified above.
 \end{nnotes}
 \vspace{-2ex}
 } %sapling
 
 
+\nufive{
+\introsection
+\lsubsubsection{\OrchardText{} Key Components}{orchardkeycomponents}
+
+Let $\PRFOutputLengthExpand$, $\SpendingKeyLength$, $\OutViewingKeyLength$, $\DiversifierLength$,
+and $\DiversifierKeyLength$ be as defined in \crossref{constants}.
+
+Let $\PRFexpand{}$ and $\PRFock{Orchard}{}$, instantiated in \crossref{concreteprfs},
+be \pseudoRandomFunctions.
+
+Let $\KA{Orchard}$, instantiated in \crossref{concreteorchardkeyagreement},
+be a \keyAgreementScheme.
+
+Let $\CommitIvk{}$, instantiated in \crossref{concretesinsemillacommit},
+be a \commitmentScheme.
+
+Let $\DiversifyHash{Orchard}$, instantiated in \crossref{concretediversifyhash},
+be a \hashFunction.
+
+Let $\SpendAuthSig{Orchard}$ instantiated in \crossref{concretespendauthsig}
+be a \rerandomizableSignatureScheme.
+
+Let $\reprP$, $\GroupP$, $\ParamP{r}$, and $\ReprP$ be as defined in \crossref{pallasandvesta}.
+and let $\GroupPHash$ be as defined in \crossref{concretegrouphashpallasandvesta}.
+
+Let $\LEBStoOSP{} \typecolon (\ell \typecolon \Nat) \times \bitseq{\ell} \rightarrow \byteseq{\sceiling{\ell/8}}$
+and $\LEOStoIP{} \typecolon (\ell \typecolon \Nat \suchthat \ell \bmod 8 = 0) \times \byteseq{\ell/8} \rightarrow \binaryrange{\ell}$
+be as defined in \crossref{endian}.
+
+Define $\ToBase{Orchard}(x \typecolon \PRFOutputExpand) := \LEOStoIPOf{\PRFOutputLengthExpand}{x} \pmod{\ParamP{q}}$.
+
+Define $\ToScalar{Orchard}(x \typecolon \PRFOutputExpand) := \LEOStoIPOf{\PRFOutputLengthExpand}{x} \pmod{\ParamP{r}}$.
+
+\introlist
+A new \Orchard \spendingKey $\SpendingKey$ is generated by choosing a bit sequence
+uniformly at random from $\SpendingKeyType$.
+
+From this \spendingKey, the \authSigningKey $\AuthSignPrivate \typecolon \GFstar{\ParamP{r}}$,
+the \authValidatingKey $\AuthSignPublic \typecolon \GroupP$,
+the \nullifierDerivingKey $\NullifierKey \typecolon \NullifierKeyTypeOrchard$,
+the \commitIvkRandomness $\CommitIvkRand \typecolon \CommitIvkRandType$,
+the \incomingViewingKey $\InViewingKey \typecolon \InViewingKeyTypeOrchard$,
+and the \outgoingViewingKey $\OutViewingKey \typecolon \OutViewingKeyType$ are derived
+as follows:
+
+\begin{algorithm}
+  \item let mutable $\AuthSignPrivate := \ToScalar{Orchard}(\PRFexpand{\SpendingKey}([6]))$
+  \item let $\NullifierKey = \ToBase{Orchard}(\PRFexpand{\SpendingKey}([7]))$
+  \item let $\CommitIvkRand = \ToScalar{Orchard}(\PRFexpand{\SpendingKey}([8]))$
+  \item if $\AuthSignPrivate = 0$, discard this key and repeat with a new $\SpendingKey$.
+  \item let mutable $\AuthSignPublic := \SpendAuthSigDerivePublic{Orchard}(\AuthSignPrivate)$
+  \item if the last bit of $\reprP(\AuthSignPublic)$ is $1$:
+  \item \tab set $\AuthSignPrivate := -\AuthSignPrivate$
+  \item \tab set $\AuthSignPublic := -\AuthSignPublic$
+  \item \blank
+  \item let $\InViewingKey = \CommitIvk{\CommitIvkRand}\big(\AuthSignPublic, \NullifierKey\big)$
```

Comment:
Should this go here?
```suggestion
  \item let $\InViewingKey = \CommitIvk{\CommitIvkRand}\big(\ExtractP\big(\AuthSignPublic\big), \NullifierKey\big)$
```

Reply:
No, the type of Commit<sup>ivk</sup> **should** be Commit<sup>ivk</sup>.Trapdoor x \mathbb{P} x F<sub>q<sub>\mathbb{P}</sub></sub> -> F<sub>r<sub>\mathbb{P}</sub></sub>.